# Guided Project: Exploring Ebay Car Sales Data
dataset: user cars from a German classified ebay called ['Kleinanzeigen'](https://www.kaggle.com/orgesleka/used-cars-database/data)

goal: clean and analyze

In [115]:
import numpy as np
import pandas as pd

autos = pd.read_csv(r'/Users/westcollins/Desktop/jupyter/guided_projects/ebayauto_project/autos.csv',encoding='Latin-1')

In [116]:
autos

dateCrawled                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        seller offerType  price   abtest vehicleType  yearOfRegistration  \
0       privat   Angebot    480     test         NaN                1993   
1       privat   Angebot  18300     test       coupe                2011   
2       privat   Angebot   9800     test         suv                2004   
3       privat   Angebot   1500     test  kleinwagen                2001   
4       privat   Angebot   3600     test  kleinwagen                2008   
...        ...       ...    ...      ...         ...                 ...   
371523  privat   Angebot   2200     test         NaN                2005   
371524  privat   Angebot   1199     test      cabrio                2000   
371525  privat   Angebot   9200     test         bus                1996   
371526  privat   Angebot   3400     test       kombi                2002   
371527  privat   Angebot  28990  control   limousine                2013   

          gearbox  powerPS        model  kilometer  monthOfRegistration  \
0         manuell        0         golf     150000                    0   
1         manuell      190          NaN     125000                    5   
2       automatik      163        grand     125000                    8   
3         manuell       75         golf     150000                    6   
4         manuell       69        fabia      90000                    7   
...           ...      ...          ...        ...                  ...   
371523        NaN        0          NaN      20000                    1   
371524  automatik      101       fortwo     125000                    3   
371525    manuell      102  transporter     150000                    3   
371526    manuell      100         golf     150000                    6   
371527    manuell      320      m_reihe      50000                    8   

       fuelType           brand notRepairedDamage          dateCreated  \
0        benzin      volkswagen               NaN  2016-03-24 00:00:00   
1        diesel            audi                ja  2016-03-24 00:00:00   
2        diesel            jeep               NaN  2016-03-14 00:00:00   
3        benzin      volkswagen              nein  2016-03-17 00:00:00   
4        diesel           skoda              nein  2016-03-31 00:00:00   
...         ...             ...               ...                  ...   
371523      NaN  sonstige_autos               NaN  2016-03-14 00:00:00   
371524   benzin           smart              nein  2016-03-05 00:00:00   
371525   diesel      volkswagen              nein  2016-03-19 00:00:00   
371526   diesel      volkswagen               NaN  2016-03-20 00:00:00   
371527   benzin             bmw              nein  2016-03-07 00:00:00   

        nrOfPictures  postalCode             lastSeen  
0                  0       70435  2016-04-07 03:16:57  
1                  0       66954  2016-04-07 01:46:50  
2                  0       90480  2016-04-05 12:47:46  
3                  0       91074  2016-03-17 17:40:17  
4                  0       60437  2016-04-06 10:17:21  
...              ...         ...                  ..

In [117]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
dateCrawled            371528 non-null object
name                   371528 non-null object
seller                 371528 non-null object
offerType              371528 non-null object
price                  371528 non-null int64
abtest                 371528 non-null object
vehicleType            333659 non-null object
yearOfRegistration     371528 non-null int64
gearbox                351319 non-null object
powerPS                371528 non-null int64
model                  351044 non-null object
kilometer              371528 non-null int64
monthOfRegistration    371528 non-null int64
fuelType               338142 non-null object
brand                  371528 non-null object
notRepairedDamage      299468 non-null object
dateCreated            371528 non-null object
nrOfPictures           371528 non-null int64
postalCode             371528 non-null int64
lastSeen              

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

    lots of German words in name column/whole data set (privat clearly private), 7 ints and 13 objs, 
    columns with nulls: vehicleType, gearbox, model, fuelType, notRepairedDamage (but no more than ~20%)
    weird amount of Crawls w/o any pictures
    Not really sure what Crawl means
    Columns are in CamelCase, but python prefers snakecase

In [118]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [119]:
autos = autos.rename(columns={'yearOfRegistration':'registration_year', 
                      'monthOfRegistration':'registration_month',
                      'notRepairedDamage':'unrepaired_damage',
                      'dateCreated':'ad_created',
                      'dateCrawled':'crawl_created',
                      'offerType':'offer_type',
                      'vehicleType':'vehicle_type',
                      'powerPS':'power_ps',
                      'fuelType':'fuel_type',
                      'nrOfPictures':'num_pictures',
                      'postalCode':'postal_code',
                      'lastSeen':'last_seen'})

the change above was made to change Camelcase to Snakecase

In [120]:
autos.head()

crawl_created                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price abtest vehicle_type  registration_year    gearbox  power_ps  model  \
0    480   test          NaN               1993    manuell         0   golf   
1  18300   test        coupe               2011    manuell       190    NaN   
2   9800   test          suv               2004  automatik       163  grand   
3   1500   test   kleinwagen               2001    manuell        75   golf   
4   3600   test   kleinwagen               2008    manuell        69  fabia   

   kilometer  registration_month fuel_type       brand unrepaired_damage  \
0     150000                   0    benzin  volkswagen               NaN   
1     125000                   5    diesel        audi                ja   
2     125000                   8    diesel        jeep               NaN   
3     150000                   6    benzin  volkswagen              nein   
4      90000                   7    diesel       skoda              nein   

            ad_created  num_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00             0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0        60437  2016-04-06 10:17:21

In [121]:
autos.describe(include='all')

crawl_created         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

        abtest vehicle_type  registration_year  gearbox       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model      kilometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  num_pictures   postal_code  \
count             299468               371528      371528.0  371528.00000   
unique                 2                  114           NaN           NaN   
top                 nein  2016-04-03 00:00:00           NaN           NaN   
freq              263182                14450           NaN           NaN   
mean                 NaN                  NaN           0.0   50820.66764   
std                  NaN                  NaN           0.0   25799.08247   
min                  NaN                  NaN           0.0    1067.00000   
25%                  NaN                  NaN           0.0   30459.00000   
50%                  NaN                  NaN           0.0   49610.00000   
75%                  NaN                  NaN           0.0   71546.00000   
max                  NaN                  NaN           0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-06 13:45:54  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In [122]:
autos['num_pictures'].value_counts()

0    371528
Name: num_pictures, dtype: int64

text where almost all data types are the same/won't be useful: (use above to see if negligible amount)
- seller
- offer_type
- num_pictures

Numeric data stored as text that can be cleaned/removed:
- supposedly price and odometer(but i think the data has been updated and fixed
- but still shown how to fix below

In [123]:
#converting column 'price' from object to numeric type
#autos['price'] = (autos['price']
                 #.str.replace('$','')
                 #.str.replace(',','')
                 #.astype(float)
                 #)

In [124]:
#converting column 'odometer' from object to numeric type
#autos['kilometer'] = (autos['kilometer']
                 #.str.replace('km','')
                 #.str.replace(',','')
                 #.astype(float)
                 #)

In [125]:
autos['price'].describe()

count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64

In [126]:
autos['price'].unique().shape

(5597,)

In [127]:
autos['price'].value_counts().head(10)

0       10778
500      5670
1500     5394
1000     4649
1200     4594
2500     4438
600      3819
3500     3792
800      3784
2000     3432
Name: price, dtype: int64

In [128]:
autos['price'].value_counts().sort_index().head(10)

0     10778
1      1189
2        12
3         8
4         1
5        26
7         3
8         9
9         8
10       84
Name: price, dtype: int64

In [129]:
autos['price'].value_counts().sort_index(ascending=False).head(10)

2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
Name: price, dtype: int64

- goes all over, 5,597 unique values is a ton. maybe could make a good graph
- too many with a value of $0, impossible
- 10 Million for a car and above seem way too high
- I think a max value around 2 million is reasonable if not a little high
- so the range should be 1 to 2,000,000

In [130]:
autos['kilometer'].describe()

count    371528.000000
mean     125618.688228
std       40112.337051
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: kilometer, dtype: float64

In [131]:
autos['kilometer'].unique().shape

(13,)

In [132]:
autos['kilometer'].value_counts()

150000    240797
125000     38067
100000     15920
90000      12523
80000      11053
70000       9773
60000       8669
50000       7615
5000        7069
40000       6376
30000       6041
20000       5676
10000       1949
Name: kilometer, dtype: int64

reasonable number of unique values here

## Exploring Date Data
There are 5 columns:
- crawl_created: added by crawler (string)
- last_seen: added by crawler (string)
- ad_created: from the website (string)
- registration_month: from the website (numerical)
- registration_year: from the website (numberical)
We can see three of the five columns are strings, so we will need to turn those into numerical columns to analyze them. Let's start by seeing how this data is formatted.

In [133]:
autos['crawl_created'].str[:10]

0         2016-03-24
1         2016-03-24
2         2016-03-14
3         2016-03-17
4         2016-03-31
             ...    
371523    2016-03-14
371524    2016-03-05
371525    2016-03-19
371526    2016-03-20
371527    2016-03-07
Name: crawl_created, Length: 371528, dtype: object

In [134]:
autos['crawl_created'].value_counts(normalize=True,dropna=False)

2016-03-24 14:49:47    0.000019
2016-03-26 22:57:31    0.000016
2016-03-19 21:49:56    0.000016
2016-03-22 10:42:10    0.000013
2016-03-23 13:50:19    0.000013
                         ...   
2016-03-10 11:48:38    0.000003
2016-03-31 09:53:20    0.000003
2016-03-25 13:58:03    0.000003
2016-03-21 17:47:52    0.000003
2016-03-08 17:52:30    0.000003
Name: crawl_created, Length: 280500, dtype: float64

In [135]:
autos['crawl_created'].sort_index()

0         2016-03-24 11:52:17
1         2016-03-24 10:58:45
2         2016-03-14 12:52:21
3         2016-03-17 16:54:04
4         2016-03-31 17:25:20
                 ...         
371523    2016-03-14 17:48:27
371524    2016-03-05 19:56:21
371525    2016-03-19 18:57:12
371526    2016-03-20 19:41:08
371527    2016-03-07 19:39:19
Name: crawl_created, Length: 371528, dtype: object

There doesn't seem to be a lot to draw from the data except that the time is military time.

In [136]:
autos['last_seen'].value_counts(normalize=True,dropna=False)

2016-04-06 13:45:54    0.000046
2016-04-07 06:45:59    0.000046
2016-04-06 09:44:33    0.000043
2016-04-07 08:15:52    0.000043
2016-04-07 09:45:10    0.000043
                         ...   
2016-03-31 06:47:23    0.000003
2016-03-24 16:17:49    0.000003
2016-04-02 23:51:42    0.000003
2016-03-24 11:16:51    0.000003
2016-03-31 18:46:34    0.000003
Name: last_seen, Length: 182806, dtype: float64

In [137]:
autos['last_seen'].sort_index()

0         2016-04-07 03:16:57
1         2016-04-07 01:46:50
2         2016-04-05 12:47:46
3         2016-03-17 17:40:17
4         2016-04-06 10:17:21
                 ...         
371523    2016-04-06 00:46:52
371524    2016-03-11 18:17:12
371525    2016-04-07 07:15:26
371526    2016-03-24 12:45:21
371527    2016-03-22 03:17:10
Name: last_seen, Length: 371528, dtype: object

In [138]:
autos['ad_created'].value_counts(normalize=True,dropna=False)

2016-04-03 00:00:00    0.038893
2016-04-04 00:00:00    0.037741
2016-03-20 00:00:00    0.036463
2016-03-12 00:00:00    0.036011
2016-03-21 00:00:00    0.035812
                         ...   
2016-01-03 00:00:00    0.000003
2015-11-10 00:00:00    0.000003
2015-12-06 00:00:00    0.000003
2015-11-02 00:00:00    0.000003
2015-06-18 00:00:00    0.000003
Name: ad_created, Length: 114, dtype: float64

In [139]:
autos['ad_created'].sort_index()

0         2016-03-24 00:00:00
1         2016-03-24 00:00:00
2         2016-03-14 00:00:00
3         2016-03-17 00:00:00
4         2016-03-31 00:00:00
                 ...         
371523    2016-03-14 00:00:00
371524    2016-03-05 00:00:00
371525    2016-03-19 00:00:00
371526    2016-03-20 00:00:00
371527    2016-03-07 00:00:00
Name: ad_created, Length: 371528, dtype: object

In [140]:
autos['registration_year'].describe()

count    371528.000000
mean       2004.577997
std          92.866598
min        1000.000000
25%        1999.000000
50%        2003.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

We can see that the minimum is 1000 and the maximum is 9999. Both are impossible values for the existence of this website. We need to fix these dates. 25% of values are less than 1999, and 75% are less than 2008. We can make the maximum year the current year so 2019. We can make the minimum value sometime in the early 90s. Let's use a time frame of 1900-2019.

In [141]:
autos['registration_year'].value_counts()

2000    24551
1999    22767
2005    22316
2006    20230
2001    20218
        ...  
1919        1
1915        1
7800        1
1400        1
8455        1
Name: registration_year, Length: 155, dtype: int64

In [142]:
autos['registration_year'].sort_values()

69429     1000
300924    1000
125299    1000
255557    1000
341569    1000
          ... 
169656    9999
52560     9999
134985    9999
210300    9999
176140    9999
Name: registration_year, Length: 371528, dtype: int64

Try head and tail here with a lot of values and you can see that 1900-2019 will work.

In [148]:
autos = autos[autos['registration_year'].between(1900,2019)]
autos['registration_year'].value_counts(normalize=True).head(20)

2000    0.066114
1999    0.061309
2005    0.060095
2006    0.054477
2001    0.054445
2003    0.053516
2004    0.053174
2002    0.051674
1998    0.048340
2007    0.047592
2008    0.043558
2009    0.042028
1997    0.039602
2010    0.033268
2011    0.032498
1996    0.029315
2017    0.028399
2016    0.026549
1995    0.026008
2012    0.025362
Name: registration_year, dtype: float64

It looks like most values are from 1995 through 2017, about a 20 year timeframe.

#### Now let's look at the brand column.

In [164]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.214366
bmw               0.108411
opel              0.108031
mercedes_benz     0.095049
audi              0.088510
ford              0.068852
renault           0.048378
peugeot           0.029689
fiat              0.026040
seat              0.018904
mazda             0.015320
skoda             0.015185
smart             0.014135
citroen           0.013947
nissan            0.013564
toyota            0.012641
sonstige_autos    0.010621
hyundai           0.009816
mini              0.009137
volvo             0.008959
mitsubishi        0.008235
honda             0.007637
kia               0.006880
alfa_romeo        0.006309
suzuki            0.006269
porsche           0.005957
chevrolet         0.004960
chrysler          0.003907
dacia             0.002424
daihatsu          0.002170
jeep              0.002170
subaru            0.002090
land_rover        0.002071
jaguar            0.001672
trabant           0.001581
daewoo            0.001460
saab              0.001425
r

We can see that volkswagen, bmw, opel, mercedes_benz, and audi lead the pack. We will choose any data with greater than a 5% share of the rows.

In [173]:
brand_counts = autos['brand'].value_counts(normalize=True)
top_brands = brand_counts[brand_counts>0.05].index
print(top_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [194]:
brand_mean = {}

for bran in top_brands:
    bran_autos = autos[autos['brand'] == bran]
    mean_price = bran_autos['price'].mean()
    brand_mean[bran] = int(mean_price)
                        
brand_mean

{'volkswagen': 14538,
 'bmw': 14797,
 'opel': 3223,
 'mercedes_benz': 17248,
 'audi': 15870,
 'ford': 8463}

Mercedes is the most expensive brand of car on average closely followed by Audi, VW, and BMW.

In [195]:
brand_mileage = {}

for bran in top_brands:
    bran_autos = autos[autos['brand'] == bran]
    mean_km = bran_autos['kilometer'].mean()
    brand_mileage[bran] = int(mean_km)
                        
brand_mileage

{'volkswagen': 128627,
 'bmw': 132813,
 'opel': 128964,
 'mercedes_benz': 130710,
 'audi': 129731,
 'ford': 123862}

Now we need to turn this dictionary into a series and then into a dataframe.

In [200]:
mean_price = pd.Series(brand_mean)
mean_mileage = pd.Series(brand_mileage)

In [207]:
aggregate_data = pd.DataFrame(mean_price, columns=['mean_price'])

In [208]:
aggregate_data['mean_mileage'] = mean_mileage
aggregate_data

mean_price  mean_mileage
volkswagen          14538        128627
bmw                 14797        132813
opel                 3223        128964
mercedes_benz       17248        130710
audi                15870        129731
ford                 8463        123862

All of the mileages are about the same, Ford has the lowest mean by a fair amount but also lies in the middle of the pack in terms of mean price. Doesn't seem to be any correlation between the data here.